Updates to experiment menu and data analysis pages

moshernoah opened the pull request:
This update corrects what the functionality of what the former "Data Analysis" button was intended for. It will renamed to "Data Exporting" and is reflected in the actual page as well. This update also creates the functionality to export files into a new "exports" folder within the Mouser folder. Each export goes into a folder sharing the name of the experiment. 

Fixes #224 

**What was changed?**

The experiment menu page button and the data analysis page itself have been renamed to Data Generation. In addition, it now takes the currently opened experiment and, with the push of a button, exports all tables into .csv file format, inserting those files into a folder sharing the experiment name inside of a new "exports" folder. 

**Why was it changed?**

This is to better align with client needs, as it allows for greater ease of external data analysis. The .csv format makes it usable with many more programs than the .mouser file protocol. 

**How was it changed?**

Currently, the titles on the button and page were edited to have the correct labels. The experiment_menu_ui page was changed to allow pertinent file information to be passed to the Data Generation Page. The Data Analysis page was also renamed to Data Generation, and it was given a button to generate the .csv exports of the currently open experiment. Lastly, the experiment_database.py was updated with several new functions. The main one, export_all_tables_to_csv, takes the currently open experiment, checks and generates an export folder, generates a folder named after the experiment inside the exports folder, and translates the database files into the several .csv files inside of the uniquely named folder.

NNewbauer commented on experiment_pages/experiment/data_analysis_ui.py:
```diff
@@ -1,10 +1,12 @@
+class DataAnalysisUI(MouserPage):
+    '''Data Exporting UI.'''
+    def __init__(self, parent: CTk, prev_page: CTkFrame = None):
+        super().__init__(parent, "Data Exporting", prev_page)
```
Standard change of naming convention in the UI. Pretty straightforward.

NNewbauer commented on experiment_pages/experiment/experiment_menu_ui.py:
```diff
@@ -42,7 +42,7 @@ def __init__(self, parent: CTk, name: str, prev_page: ChangeableFrame = None, co
         self.collection_button = CTkButton(main_frame, text='Data Collection', width=button_size,
                                 command= self.data_page.raise_frame)
-        self.analysis_button = CTkButton(main_frame, text='Data Analysis', width=button_size,
+        self.analysis_button = CTkButton(main_frame, text='Data Exporting', width=button_size,
```
Again, just a change of naming convention for UI. Let me know once the meat of the code is pushed and I'll report back.

NNewbauer reviewed (commented):
Quick comments, I'll make more once more is pushed.

moshernoah commented:
All changes have been made. Ready for re-review @NNewbauer @jackbelyeu

moshernoah commented:
Note: One of the files had imports added to it so github thinks I deleted and rewrote the entire document. I did not. I will comment where the changes at the top of the file are, but the real ones come from the last two methods of that file.

moshernoah commented on databases/experiment_database.py:
```diff
@@ -1,445 +1,512 @@
+import os
+import csv
+from customtkinter import CTk, CTkLabel, CTkButton
+
```
These imports were an actual change by me.

moshernoah commented on databases/experiment_database.py:
```diff
@@ -1,445 +1,512 @@
+class ExperimentDatabase:
+    '''SQLite Database Object for Experiments.'''
+    def __init__(self, file=":memory:"):  #call with file name as argument or no args to use memory
+        self.db_file = file
```
This line is the only thing in the first 400 lines that I actually changed. But it pushed everything else down in line numbers

moshernoah commented on databases/experiment_database.py:
```diff
@@ -1,445 +1,512 @@
+        '''Closes database file.'''
+        self._conn.close()
+
+    def export_all_tables_to_csv(self, export_dir="exports"):
```
This line is where my changes to the code actually begin.

kate-holdener commented on databases/experiment_database.py:
```diff
@@ -1,445 +1,512 @@
+import sqlite3
+import os
+import csv
+from customtkinter import CTk, CTkLabel, CTkButton
```
I am concerned about having database related code depend on UI elements. Can you think of a way to remove the UI dependency here?

kate-holdener commented on databases/experiment_database.py:
```diff
@@ -1,445 +1,512 @@
+        print("All tables have been exported successfully.")
+
+        # Once export is done, show the success notification
+        self.show_success_notification()
```
Perhaps this function could return a success status, and the caller of export_all_tables_to_csv could show the success notification. This would allow you to remove the UI dependency from the database code.

jackbelyeu replied:
Here are two ideas:
As Dr. Holdener suggested, export_all_tables_to_csv should really just perform a single function -- exporting tables to csv. The UI logic should be handled in /experiment_pages/experiment/data_analysis_ui.py when export_to_csv is called. If you anticipate this UI element (the successful export popup) to be needed across multiple pages, then we should look into additional abstraction.

moshernoah commented:
> Please remove the customtkinter dependency from experiment_database.py (see my comments for details).

This change has been addressed. All UI elements have been moved from experiment_database.py and added instead to data_analysis_ui.py and the customtkinter dependency has been removed.

kate-holdener commented on experiment_pages/experiment/data_analysis_ui.py:
```diff
@@ -1,10 +1,56 @@
+            # Once export is done, show the success notification
+            self.show_success_notification()
+        except Exception as e:
+            print(f"An error occurred while exporting data: {e}")
```
Would it be better to show the error through the UI, just like you show success notification?

jackbelyeu replied:
I've gone ahead and added this. I'd be nice if the UI was more specific with why it failed, but that is another issue.

jackbelyeu commented on databases/experiment_database.py:
```diff
@@ -1,445 +1,497 @@
+            os.makedirs(export_dir)
+
+        # Get the base name of the database file (without extension)
+        db_name = os.path.splitext(os.path.basename(self.db_file))[0]
```
The branch does successfully create the .csv files in my testing, under normal experiments and application usage. I'm not sure about the edge case you're referencing.

kate-holdener replied:
The constructor for ExperimentDatabase defaults filename to ":memory:" if no argument for file is provided. Not sure if that's used anywhere or not.

kate-holdener replied:
I'd check to see if the ExperiementDatabase is ever created without a file argument? If not, maybe it's best to remove the default ":memory:" from the constructor. That might be the simplest solution, perhaps that feature is just no longer used.

moshernoah replied:
It looks like .mouser file creation and editing is unaffected by the removal of the default constructor. To my testing, it can be safely removed.

kate-holdener commented on experiment_pages/experiment/data_analysis_ui.py:
```diff
@@ -1,10 +1,71 @@
+        '''Raise the frame for this UI'''
+        super().raise_frame()
+
+    def show_success_notification(self):
+        '''Displays a success notification when export is complete.'''
+        notification = CTk()  # Create a new window
+        notification.title("Export Successful")
+        notification.geometry("300x100")  # Set the window size
+
+        label = CTkLabel(notification, text="Export to CSV was successful!", pady=20)
+        label.pack()
+
+        ok_button = CTkButton(notification, text="OK", command=notification.destroy)
+        ok_button.pack(pady=10)
+
+        notification.mainloop()
+
+    def show_failure_notification(self):
+        '''Displays a failure notification when export is complete.'''
+        notification = CTk()  # Create a new window
+        notification.title("Export Failed")
+        notification.geometry("300x100")  # Set the window size
+
+        label = CTkLabel(notification, text="Export to CSV failed.", pady=20)
+        label.pack()
+
+        ok_button = CTkButton(notification, text="OK", command=notification.destroy)
+        ok_button.pack(pady=10)
+
+        notification.mainloop()
```
You can probably generalize this code to avoid code duplication. Just send the title and text as parameters to a show_notification function.

```suggestion
    def show_notification(self, title, message):
        '''Displays a success notification when export is complete.'''
        notification = CTk()  # Create a new window
        notification.title(title)
        notification.geometry("300x100")  # Set the window size

        label = CTkLabel(notification, text=message, pady=20)
        label.pack()

        ok_button = CTkButton(notification, text="OK", command=notification.destroy)
        ok_button.pack(pady=10)

        notification.mainloop()

```

Then call show_notification instead of calling show_sucess_notificaton and show_failure_notification, passing appropriate values for title and message.

moshernoah commented:
@kate-holdener Would you remove your merge block? The single notification method has been updated and the UI customtkinter elements have been removed from experiment_database.py.
